src/sc_crawler/vendors/azure.py

```diff
@@ -0,0 +1,972 @@
+    except HttpResponseError:
+        pass
+    return servers
+
+
+@cache
+def _prices(url_params: Optional[str] = None) -> List[dict]:
+    session = request_session()
+    data = []
+    next_url = "https://prices.azure.com/api/retail/prices"
+    if url_params:
+        next_url += "?" + url_params
+    while next_url:
+        response = session.get(next_url)
+        # handle rate limiting
+        headers = response.headers
+        remaining = headers.get("x-ms-ratelimit-remaining-retailPrices-requests", 0)
+        if int(remaining) == 0:
+            logger.debug("Retail Prices API rate limit reached, sleep for 60 seconds.")
+            sleep(60)
+        # next page
+        if response.status_code == 200:
+            json = response.json()
+            next_url = json.get("NextPageLink")
+            data += json["Items"]
+        else:
+            raise HttpResponseError(response.content)
+    return data
+
+
+# ##############################################################################
+# Internal helpers
+
+
+SERVER_FEATURES = {
+    # a = AMD-based processor
+    # b = Block Storage performance
+    # d = diskful (that is, a local temp disk is present)
+    # i = isolated size
+    # l = low memory; a lower amount of memory than the memory intensive size
+    # m = memory intensive; the most amount of memory in a particular size
+    # p = ARM Cpu
+    # t = tiny memory; the smallest amount of memory in a particular size
+    # s = Premium Storage capable, including possible use of Ultra SSD
+    "a": "AMD processor",
+    "p": "ARM processor",
+    "b": "Block Storage performance",
+    "d": "Local Disk",
+    "i": "Isolated",
+    "l": "Low Memory",
+    "m": "Memory Intensive",
+    "t": "Tiny Memory",
+    "s": "Premium Storage capable",
+    # not part of the official docs, searched at randomg places
+    "r": "RDMA capable",
+    "e": "Memory Optimized",  # Standard_DC2es_v5
+}
+"""Map lowercase chars from the server name to features."""
+
+STORAGE_METER_MAPPING = {
+    "P2 LRS Disk Mount": ("PremiumV2_LRS", 1),
+    "P1 LRS Disk": ("Premium_LRS", 4),
+    "P1 ZRS Disk": ("Premium_ZRS", 4),
+    "E1 LRS Disk": ("StandardSSD_LRS", 4),
+    "E1 ZRS Disk": ("StandardSSD_ZRS", 4),
+    "S4 LRS Disk": ("Standard_LRS", 32),
+    "Ultra LRS Provisioned Capacity": ("UltraSSD_LRS", 1),
+}
+"""Map Storage price meter names to the Storage name and the related disk's size."""
+
+
+def _parse_server_name(name):
+    """Extract information from the server name/size.
+
+    Based on <https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/overview>."""
+    # drop the constant prefix
+    name = name.removeprefix("Standard_")
+    # first ALLCAPS chars are the family name (we don't care about the subfamily for now)
+    family = search(r"^([A-Z]*)", name).group(1)
+    name = name.removeprefix(family)
+    vcpus = int(search(r"^([0-9]*)", name).group(1))
+    name = name.removeprefix(str(vcpus))
```

Comment:
Instead of this, wouldn't something like the following be better?
```python
def parse_azure_instance_name(instance_name):
    # Updated regular expression to handle optional constrained vCPUs
    pattern = re.compile(
        r'(?P<family>[A-Za-z]+)_'
        r'(?P<subfamily>[A-Za-z])'
        r'(?P<vcpu>\d+)(?:-\d+)?'  # Match optional constrained vCPUs
        r'(?P<feature1>[a-z]?)'
        r'(?P<feature2>[a-z]?)'
        r'(?P<feature3>[a-z]?)'
        r'(_(?P<accelerator>[a-zA-Z0-9]+))?'  # Capture optional accelerator
        r'(_v(?P<version>\d+))$'              # Capture mandatory version, ensure it is the last part
    )

    match = pattern.match(instance_name)
    if not match:
        raise ValueError(f"Instance name '{instance_name}' does not match the expected format.")

    parsed_data = match.groupdict()

    # Determine architecture based on feature1
    feature1 = parsed_data['feature1']
    if feature1 == 'p':
        architecture = 'arm64'
    else:
        architecture = 'x86_64'

    # Create a set of features, filtering out empty strings and None
    features = {parsed_data['feature1'], parsed_data['feature2'], parsed_data['feature3']}
    features.discard('')  # Remove empty strings from the set if present
    features.discard(None)  # Remove None from the set if present

    # Extract the accelerator if present
    accelerator = parsed_data.get('accelerator')

    # Convert vCPU to integer
    parsed_data['vcpu'] = int(parsed_data['vcpu'])

    # Add the architecture, features set, and accelerator to the parsed data
    parsed_data['architecture'] = architecture
    parsed_data['features'] = features
    parsed_data['accelerator'] = accelerator

    # Add the version directly
    parsed_data['version'] = parsed_data.pop('version')

    # Optionally remove individual feature keys if not needed
    parsed_data.pop('feature1')
    parsed_data.pop('feature2')
    parsed_data.pop('feature3')

    return parsed_data
```

Ah, it needs further tweaking according to the list on https://cloudprice.net/.
For example it says that if there's an `N` in the name, it's GPU enabled. And a lot more.

This is a nightmare. :)

Reply:
> For example it says that if there's an N in the name, it's GPU enabled. And a lot more.

See e.g. the NP family, which comes with an FPGA, not a GPU :shrug: 
https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/fpga-accelerated/np-family

I'd rather stick with the official docs, which is not fun, but hopefully reliable.

Regarding the subfamily, as per my understanding: that's the "P" from "NP" -- but I did not want to spend too much time on it, as it's not relevant in our schema .. so just kept the family + subfamily altogether for our use case (that is BTW more in sync with the AWS/GCP/Hetzner instance family sizes).

I also did't want to count the max number of feature flags :) But after a quick search, there's e.g. `Standard_E112ibds_v5` with 4.

Overall, I tried to keep this as relaxed as possible, but we might burn ourselves with it ... let's sleep on this.